dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx

```diff
@@ -190,31 +190,65 @@ cpdef utils.dpnp_descriptor dpnp_identity(n, result_dtype):
     return result
 
 
-# TODO this function should work through dpnp_arange_c
-cpdef tuple dpnp_linspace(start, stop, num, endpoint, retstep, dtype, axis):
-    cdef shape_type_c obj_shape = utils._object_to_tuple(num)
-    cdef utils.dpnp_descriptor result = utils_py.create_output_descriptor_py(obj_shape, dtype, None)
+cpdef utils.dpnp_descriptor dpnp_linspace(start, stop, num, dtype=None, device=None, usm_type=None, sycl_queue=None, endpoint=True, retstep=False, axis=0):
+    usm_type_alloc, sycl_queue_alloc = utils_py.get_usm_allocations([start, stop])
 
-    if endpoint:
-        steps_count = num - 1
-    else:
-        steps_count = num
+    if sycl_queue is None and device is None:
+        sycl_queue = sycl_queue_alloc
+    sycl_queue_normalized = dpnp.get_normalized_queue_device(sycl_queue=sycl_queue, device=device)
 
-    # if there are steps, then fill values
-    if steps_count > 0:
-        step = (dpnp.float64(stop) - start) / steps_count
-        for i in range(1, result.size):
-            result.get_pyobj()[i] = start + step * i
+    if usm_type is None:
+        usm_type = "device" if usm_type_alloc is None else usm_type_alloc
+
+    if isinstance(start, (list, tuple)):
+        start = dpnp.asarray(start, usm_type=usm_type, sycl_queue=sycl_queue_normalized)
+
+    if isinstance(stop, (list, tuple)):
+        stop = dpnp.asarray(stop, usm_type=usm_type, sycl_queue=sycl_queue_normalized)
+
+    dt = numpy.result_type(start, stop, float(num))
+
+    dt = utils_py.map_dtype_to_device(dt, sycl_queue_normalized.sycl_device)
+
+    if dtype is None:
+        dtype = dt
+
+    if dpnp.isscalar(start) and dpnp.isscalar(stop):
+        res = dpnp_container.linspace(start,
+                                      stop,
+                                      num,
+                                      dtype=dt,
+                                      device=device,
+                                      usm_type=usm_type,
+                                      sycl_queue=sycl_queue_normalized,
+                                      endpoint=endpoint)
     else:
-        step = dpnp.nan
+        num = operator.index(num)
+        if num < 0:
+            raise ValueError("Number of points must be non-negative")
 
-    # if result is not empty, then fiil first and last elements
-    if num > 0:
-        result.get_pyobj()[0] = start
-        if endpoint and result.size > 1:
-            result.get_pyobj()[result.size - 1] = stop
+        _start = dpnp.asarray(start, dtype=dt, usm_type=usm_type, sycl_queue=sycl_queue_normalized)
+        _stop = dpnp.asarray(stop, dtype=dt, usm_type=usm_type, sycl_queue=sycl_queue_normalized)
```

Comment:
In case `start` and `stop` are both `dpnp_array` or `usm_ndarray`, we might change `usm_type` type here for resulting `_start` and `_stop` and so zero will not be applied (i.e. extra memory copy might take place).